akka-http-core/src/main/scala/akka/http/impl/engine/client/pool/HostConnectionPool.scala

```diff
@@ -0,0 +1,335 @@
+
+        setHandlers(requestsIn, responsesOut, this)
+
+        val slots = Vector.tabulate(settings.maxConnections)(new Slot(_))
+        var outBuffer: util.Deque[ResponseContext] = new util.ArrayDeque[ResponseContext]
+        var retryBuffer: util.Deque[RequestContext] = new util.ArrayDeque[RequestContext]
+
+        // TODO: implement min-connections
+        override def preStart(): Unit = {
+          pull(requestsIn)
+          slots.foreach(_.initialize())
+        }
+
+        def onPush(): Unit = {
+          dispatchRequest(grab(requestsIn))
+          pullIfNeeded()
+        }
+        def onPull(): Unit =
+          if (!outBuffer.isEmpty)
+            push(responsesOut, outBuffer.pollFirst())
+
+        def manageState(): Unit = {
+          pullIfNeeded()
+        }
+
+        def pullIfNeeded(): Unit =
+          if (hasIdleSlots)
+            if (!retryBuffer.isEmpty) {
+              log.debug("Dispatching request from retryBuffer")
+              dispatchRequest(retryBuffer.pollFirst())
+            } else if (!hasBeenPulled(requestsIn))
+              pull(requestsIn)
+
+        def hasIdleSlots: Boolean =
+          // FIXME: optimize by keeping track of idle connections
+          slots.exists(_.isIdle)
+
+        def dispatchResponse(req: RequestContext, res: HttpResponse): Unit =
+          dispatchResponseContext(ResponseContext(req, Success(res)))
+
+        def dispatchFailure(req: RequestContext, cause: Throwable): Unit = {
+          // FIXME: or retry
+          if (req.retriesLeft > 0) {
+            log.debug("Request has {} retries left, retrying...", req)
+            retryBuffer.addLast(req.copy(retriesLeft = req.retriesLeft - 1))
+          } else
+            dispatchResponseContext(ResponseContext(req, Failure(cause)))
+        }
+
+        def dispatchResponseContext(resCtx: ResponseContext): Unit =
+          if (outBuffer.isEmpty && isAvailable(responsesOut))
+            push(responsesOut, resCtx)
+          else
+            outBuffer.addLast(resCtx)
+
+        def dispatchRequest(req: RequestContext): Unit = {
+          val slot =
+            slots.find(_.isIdle)
+              .getOrElse(throw new IllegalStateException("Tried to dispatch request when no slot is idle"))
+
+          slot.debug("Dispatching request") // FIXME: add abbreviation
+          slot.dispatchRequest(req)
+        }
+
+        private class Slot(val slotId: Int) extends SlotContext {
+          private[this] var state: SlotState = SlotState.Unconnected
+          private[this] var connection: SlotConnection = _
+          def isIdle: Boolean = state.isIdle
+          def isConnected: Boolean = state.isConnected
+          def shutdown(): Unit = {
+            // TODO: should we offer errors to the connection?
+            closeConnection()
+
+            state.onShutdown(this)
+          }
+
+          def initialize(): Unit =
+            if (slotId < settings.minConnections)
+              updateState(_.onPreConnect(this))
```

Comment:
Is the above `+        // TODO: implement min-connections` not needed anymore? Looks like that's the opening connections eagerly already as expected